File: 0142.linked_list_cycle_II/solution.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def detect_cycle(self, head: ListNode):
        fast, slow = head, head
        while fast and fast.next:
            fast, slow = fast.next.next, slow.next
            if fast == slow:
                break
        if not(fast and fast.next):
            return
        while head != slow:
            head, slow = head.next, slow.next
        return head

    def hash_solution(self, head: ListNode):
        if not head:
            return None
        node = set()
        while head:
            if head in node:
                return head
            else:
                node.add(head)
            head = head.next

File: 0142.linked_list_cycle_II/solution.go
package leetcode0142

import "github.com/Ackerr/Algorithms/utils"

func detectCycle(head *utils.ListNode) *utils.ListNode {
	fast, slow := head, head
	// Check is cycle linked list
	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
		if fast == slow {
			break
		}
	}
	if slow == nil || slow.Next == nil {
		return nil
	}
	for head != slow {
		head, slow = head.Next, slow.Next
	}
	return head
}
